README.md
# Credit Card Slider - Flutter

This project provides a credit card slider

Author: theboringdeveloper

[Medium](https://medium.com/@theboringdeveloper)

[Instagram](https://www.instagram.com/theboringdeveloper/)

## Screenshots

![Demo 1](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/screenshot_1.gif)

## Creating Credit Cards

### Basic Credit card

    CreditCard(
        cardBackground: SolidColorCardBackground(Colors.red),
        cardNetworkType: CardNetworkType.visaBasic,
        cardHolderName: 'The boring developer',
    ),

![](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/basic_card.png)

* Creates a credit card with a solid background of red color
* Credit card of type visa
* Shows a card holder name

## Creating Card Slider

    @override
    Widget build(BuildContext context) {
        return Scaffold(
          body: Center(
            child: CreditCardSlider(
              _creditCards, //List of credit cards
            ),
          ),
        );
    }

### Changing percent of upper card shown

Use field ***percentOfUpperCard*** of **CreditCardSlider**

**NOTE:** Value must lie between 0 and pi / 2

    @override
    Widget build(BuildContext context) {
        return Scaffold(
          body: Center(
            child: CreditCardSlider(
              _creditCards,
              percentOfUpperCard: pi / 2,
            ),
          ),
        );
    }

![Demo 1](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/screenshot_2.gif)

## Specify initial card
**NOTE:** Added in 1.0.1

You can also specify initial card index to appear in slider

* initialCard is 0 index based
* If not specified, initialCard is 0

      @override
      Widget build(BuildContext context) {
        return Scaffold(
          body: Center(
            child: CreditCardSlider(
              _creditCards,
              initialCard: 2,
            ),
          ),
        );
      }

## Repeat cards
**NOTE:** Added in 1.0.1

You can also repeat cards in down or both direction (up and down)

### Repeat in down direction only

      @override
      Widget build(BuildContext context) {
        return Scaffold(
          body: Center(
            child: CreditCardSlider(
              _creditCards,
              repeatCards: RepeatCards.down,
            ),
          ),
        );
      }

![Demo 1](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/screenshot_3.gif)

### Repeat in both direction

      @override
      Widget build(BuildContext context) {
        return Scaffold(
          body: Center(
            child: CreditCardSlider(
              _creditCards,
              repeatCards: RepeatCards.bothDirection,
            ),
          ),
        );
      }

![Demo 1](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/screenshot_4.gif)


## Card click listener

    @override
    Widget build(BuildContext context) {
        return Scaffold(
            body: Center(
                child: CreditCardSlider(
                    _creditCards,
                    onCardClicked: (index) {
                        print('Clicked at index: $index');
                    },
                ),
            ),
        );
    }

* Added in 1.0.0+2

### Card Background Types

* Solid Color
* Gradient
* Image

### Credit card with Gradient background

    CreditCard(
        cardBackground: GradientCardBackground(LinearGradient(
          begin: Alignment.centerLeft,
          end: Alignment.centerRight,
          colors: [Color(0xFF4AA3F2), Color(0xFFAF92FB)],
          stops: [0.3, 0.95],
        )),
        cardNetworkType: CardNetworkType.visaBasic,
        cardHolderName: 'The boring developer',
    )

![Gradient card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/gradient_card.png)

### Credit card with Image background

    CreditCard(
        cardBackground:
            ImageCardBackground(AssetImage('images/background_sample.jpg')),
        cardNetworkType: CardNetworkType.rupay,
        cardHolderName: 'The boring developer',
    )

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/image_card.png)

### Adding Credit card number

***cardNumber*** field of type string

    CreditCard(
        cardBackground: SolidColorCardBackground(Colors.black.withOpacity(0.6)),
        cardNetworkType: CardNetworkType.visaBasic,
        cardHolderName: 'The boring developer',
        cardNumber: '1234 1234 1234 1234',
    )

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/credit_card_with_number.png)

    CreditCard(
        cardBackground: SolidColorCardBackground(Colors.black.withOpacity(0.6)),
        cardNetworkType: CardNetworkType.visaBasic,
        cardHolderName: 'The boring developer',
        cardNumber: '1234 **** **** ****',
    )

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/credit_card_with_hidden_numbers.png)

### Adding a credit card company

    CreditCard(
        cardBackground: SolidColorCardBackground(Colors.black.withOpacity(0.6)),
        cardNetworkType: CardNetworkType.visaBasic,
        cardHolderName: 'The boring developer',
        cardNumber: '1234 **** **** ****',
        company: CardCompany.hdfc,
    )

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/hdfc_credit_card.png)

### Creating a custom company card

    CreditCard(
        cardBackground: SolidColorCardBackground(Colors.black.withOpacity(0.6)),
        cardNetworkType: CardNetworkType.visaBasic,
        cardHolderName: 'The boring developer',
        cardNumber: '1234 **** **** ****',
        company: CardCompany(Image.asset(imagePath)),
    )

**NOTE:** - List of companies available at end

### Adding Valid Till

**NOTE:** If Validity is added, valid thru must be added. Valid from is optional

      CreditCard(
        cardBackground: SolidColorCardBackground(Colors.purple),
        cardNetworkType: CardNetworkType.visaBasic,
        cardHolderName: 'The boring developer',
        cardNumber: '1234 **** **** ****',
        company: CardCompany.sbi,
        validity: Validity(
          validThruMonth: 1,
          validThruYear: 21,      
        ),
      ),

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/valid_till_card.png)

### Adding Valid From and Valid Till

      CreditCard(
        cardBackground: SolidColorCardBackground(Colors.purple),
        cardNetworkType: CardNetworkType.visaBasic,
        cardHolderName: 'The boring developer',
        cardNumber: '1234 **** **** ****',
        company: CardCompany.sbi,
        validity: Validity(
          validThruMonth: 1,
          validThruYear: 21,
          validFromMonth: 1,
          validFromYear: 16,
        ),
      ),

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/valid_from_card.png)

### Hiding chip

Set ***showChip*** to false

      CreditCard(
        cardBackground: SolidColorCardBackground(Colors.purple),
        cardNetworkType: CardNetworkType.visaBasic,
        cardHolderName: 'The boring developer',
        cardNumber: '1234 **** **** ****',
        company: CardCompany.sbi,
        showChip: false,
        validity: Validity(
          validThruMonth: 1,
          validThruYear: 21,
          validFromMonth: 1,
          validFromYear: 16,
        ),
      ),

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/hiding_chip.png)

### Changing Text color

      CreditCard(
        cardBackground: SolidColorCardBackground(Colors.white),
        cardNetworkType: CardNetworkType.visaBasic,
        cardHolderName: 'The boring developer',
        cardNumber: '1234 **** **** ****',
        company: CardCompany.sbi,
        validity: Validity(
          validThruMonth: 1,
          validThruYear: 21,
          validFromMonth: 1,
          validFromYear: 16,
        ),
        numberColor: Colors.black,
        validityColor: Colors.black,
        cardHolderNameColor: Colors.black,
      ),

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/white_card_black_text.png)

### List of companies

* American Express

        company: CardCompany.americanExpress

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/american_express.png)

* AxisBank

        company: CardCompany.axisBank

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/axis_bank_card.png)

        company: CardCompany.axisBankWhite

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/axis_bank_white_card.png)

* Citi Bank

        company: CardCompany.citiBank

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/citi_bank_card.png)

* HDFC

* HSBC

        company: CardCompany.hsbc

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/hsbc_card.png)

* ICICI

        company: CardCompany.icici

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/icici_card.png)

* INDUSLAND

        company: CardCompany.indusland

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/indusland_card.png)

* Kotak

        company: CardCompany.kotak

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/kotak_card.png)

* SBI

        company: CardCompany.sbi

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/sbi_card.png)

* Virgin

        company: CardCompany.virgin

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/virgin_card.png)

* YESBANK

        company: CardCompany.yesBank

![Image card](https://github.com/GursheeshSingh/creditcard_slider/raw/master/screenshots/yes_bank_card.png)



CHANGELOG.md
## 2.0.0 - 24 February 2021
* Migrated package to null safety

## 1.0.0 - 27 May 2020
* Basic card slider

## 1.0.0+1 - 27 May 2020
* Updating Description

## 1.0.0+2 - 27 May 2020
* On card click listener added

## 1.0.0+3 - 3 June 2020
* Fixing README images

## 1.0.0+4 - 3 June 2020
* Fixing README images

## 1.0.1 - 10 July 2020
* Repeat cards added
* Specify initial card added
* Fixed display when loading cards for first time


example/lib/main.dart
import 'package:credit_card_slider/card_background.dart';
import 'package:credit_card_slider/card_company.dart';
import 'package:credit_card_slider/card_network_type.dart';
import 'package:credit_card_slider/credit_card_slider.dart';
import 'package:credit_card_slider/credit_card_widget.dart';
import 'package:credit_card_slider/validity.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MainPage(),
    );
  }
}

class MainPage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        padding: EdgeInsets.symmetric(horizontal: 24),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            ElevatedButton(
              style: ElevatedButton.styleFrom(
                primary: kBlue
              ),
              child: Text('Simple card slider',
                  style: TextStyle(color: Colors.white)),
              onPressed: () {
                Navigator.push(
                  context,
                  MaterialPageRoute(
                    builder: (context) => SimpleCreditCardSlider(),
                  ),
                );
              },
            ),
            ElevatedButton(
              style: ElevatedButton.styleFrom(
                  primary: kBlue
              ),
              child: Text('Credit card slider with repeating down cards',
                  style: TextStyle(color: Colors.white)),
              onPressed: () {
                Navigator.push(
                  context,
                  MaterialPageRoute(
                    builder: (context) => CreditCardSliderRepeatDown(),
                  ),
                );
              },
            ),
            ElevatedButton(
              style: ElevatedButton.styleFrom(
                primary: kBlue,
                padding: const EdgeInsets.all(8),
              ),
              child: Text(
                  'Credit card slider with repeating cards in both direction',
                  style: TextStyle(color: Colors.white)),
              onPressed: () {
                Navigator.push(
                  context,
                  MaterialPageRoute(
                    builder: (context) => CreditCardSliderRepeatBothSides(),
                  ),
                );
              },
            ),
          ],
        ),
      ),
    );
  }
}

Color kPink = Color(0xFFEE4CBF);
Color kRed = Color(0xFFFA3754);
Color kBlue = Color(0xFF4AA3F2);
Color kPurple = Color(0xFFAF92FB);

var _creditCards = [
  CreditCard(
    cardBackground: SolidColorCardBackground(Colors.black.withOpacity(0.6)),
    cardNetworkType: CardNetworkType.visaBasic,
    cardHolderName: 'The boring developer',
    cardNumber: '1234 1234 1234 1234',
    company: CardCompany.yesBank,
    validity: Validity(
      validThruMonth: 1,
      validThruYear: 21,
      validFromMonth: 1,
      validFromYear: 16,
    ),
  ),
  CreditCard(
    cardBackground: SolidColorCardBackground(kRed.withOpacity(0.4)),
    cardNetworkType: CardNetworkType.mastercard,
    cardHolderName: 'Gursheesh Singh',
    cardNumber: '2434 2434 **** ****',
    company: CardCompany.kotak,
    validity: Validity(
      validThruMonth: 1,
      validThruYear: 21,
    ),
  ),
  CreditCard(
    cardBackground: GradientCardBackground(LinearGradient(
      begin: Alignment.centerLeft,
      end: Alignment.centerRight,
      colors: [kBlue, kPurple],
      stops: [0.3, 0.95],
    )),
    cardNetworkType: CardNetworkType.mastercard,
    cardHolderName: 'Very Very very boring devloper',
    cardNumber: '4567',
    company: CardCompany.sbiCard,
    validity: Validity(
      validThruMonth: 2,
      validThruYear: 2021,
    ),
  ),
  CreditCard(
    cardBackground:
        ImageCardBackground(AssetImage('images/background_sample.jpg')),
    cardNetworkType: CardNetworkType.mastercard,
    cardHolderName: 'John Doe',
    cardNumber: '2434 2434 **** ****',
    company: CardCompany.virgin,
    validity: Validity(
      validThruMonth: 1,
      validThruYear: 20,
    ),
  ),
  CreditCard(
    cardBackground:
    ImageCardBackground(AssetImage('images/background_sample.jpg')),
    cardNetworkType: CardNetworkType.americanExpress,
    cardHolderName: 'John Doe',
    cardNumber: '2434 2434 **** ****',
    company: CardCompany.virgin,
    validity: Validity(
      validThruMonth: 1,
      validThruYear: 20,
    ),
  ),
  CreditCard(
    cardBackground:
    ImageCardBackground(AssetImage('images/background_sample.jpg')),
    cardNetworkType: CardNetworkType.discover,
    cardHolderName: 'John Doe',
    cardNumber: '2434 2434 **** ****',
    company: CardCompany.virgin,
    validity: Validity(
      validThruMonth: 1,
      validThruYear: 20,
    ),
  ),
  CreditCard(
    cardBackground:
    ImageCardBackground(AssetImage('images/background_sample.jpg')),
    cardNetworkType: CardNetworkType.unionPay,
    cardHolderName: 'John Doe',
    cardNumber: '2434 2434 **** ****',
    company: CardCompany.virgin,
    validity: Validity(
      validThruMonth: 1,
      validThruYear: 20,
    ),
  ),
  CreditCard(
    cardNetworkType: CardNetworkType.rupay,
    cardHolderName: 'THE BORING DEVELOPER',
    cardNumber: '2434 2434 **** ****',
    company: CardCompany.sbi,
  ),
];

//Opens a credit card slider with initial card as 2

class SimpleCreditCardSlider extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: CreditCardSlider(
          _creditCards,
          initialCard: 2,
          onCardClicked: (index) {
            print('Clicked at index: $index');
          },
        ),
      ),
    );
  }
}

class CreditCardSliderRepeatDown extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: CreditCardSlider(
          _creditCards,
          repeatCards: RepeatCards.down,
        ),
      ),
    );
  }
}

class CreditCardSliderRepeatBothSides extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: CreditCardSlider(
          _creditCards,
          repeatCards: RepeatCards.bothDirection,
        ),
      ),
    );
  }
}
